### PR TITLE
Provides an ErrorContext which permits to create more concise error m…

### PR DESCRIPTION
**Pull request by andyHa**
…essages.

### CODE REVIEW

**Review by meg1502 — APPROVED** [review]

**Review by jmuscireum — APPROVED** [review]

File: src/main/java/sirius/biz/process/ErrorContext.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+     * @param value the value to store
+     * @param task  the task to perform while the value is set
+     * @return the context for fluent method calls
+     * @throws Exception any error as thrown by the given task. Note that the exceptions and not handlend and also
+     *                   not automatically {@link #enhanceMessage(String) enhanced}.
+     */
+    public ErrorContext inContext(String label, Object value, UnitOfWork task) throws Exception {
+        withContext(label, value);
+        task.execute();
+        removeContext(label);
```
Comment (qw3ry) [inline src/main/java/sirius/biz/process/ErrorContext.java:96]: `UnitOfWork` _does_ throw, right? Maybe put a `finally` here?

File: src/main/java/sirius/biz/process/ErrorContext.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    public void perform(UnaryOperator<String> failureDescription, UnitOfWork task) {
+        try {
+            task.execute();
+        } catch (HandledException exception) {
+            if (exception.getHint(MESSAGE_ENHANCED).asBoolean()) {
+                throw exception;
+            }
+
+            throw Exceptions.createHandled()
+                            .error((exception.getCause() instanceof HandledException) ? null : exception.getCause())
```
Comment (qw3ry) [inline src/main/java/sirius/biz/process/ErrorContext.java:130]: hm. Why not always add the catched exception as the cause?

File: src/main/java/sirius/biz/jobs/batch/file/DictionaryBasedImport.java [hunk 3]
```diff
@@ -138,32 +122,17 @@ private void handleDataRow(int index, Values row) {
...
-                                     .handle());
-        } finally {
+        Context context = filterEmptyValues(dictionary.load(row, false));
+        if (!isEmptyContext(context)) {
+            Watch watch = Watch.start();
+            errorContext.perform(() -> {
+                try {
+                    rowHandler.invoke(Tuple.create(index, context));
+                } catch (Exception exception) {
+                    process.incCounter(NLS.get("LineBasedJob.erroneousRow"));
```
Comment (idlira) [inline src/main/java/sirius/biz/jobs/batch/file/DictionaryBasedImport.java:132]:
```suggestion
                    process.incCounter("LineBasedJob.erroneousRow");
```

File: src/main/java/sirius/biz/jobs/batch/file/EntityImportJob.java [hunk 4]
```diff
@@ -106,7 +104,7 @@ protected final void handleRow(int index, Context context) {
         }
 
         E entity = findAndLoad(context);
-        try {
+        errorContext.performImport(entity::toString, () -> {
             if (shouldSkip(entity)) {
                 process.incCounter(NLS.get("EntityImportJob.rowIgnored"));
```
Comment (idlira) [inline src/main/java/sirius/biz/jobs/batch/file/EntityImportJob.java:109]:
```suggestion
                process.incCounter("EntityImportJob.rowIgnored");
```

File: src/main/java/sirius/biz/jobs/batch/file/EntityImportJob.java [hunk 5]
```diff
@@ -126,13 +124,7 @@ protected final void handleRow(int index, Context context) {
             } else {
                 process.addTiming(NLS.get("EntityImportJob.entityUpdated"), watch.elapsedMillis());
```
Comment (idlira) [inline src/main/java/sirius/biz/jobs/batch/file/EntityImportJob.java:125]:
```suggestion
                process.addTiming("EntityImportJob.entityUpdated", watch.elapsedMillis());
```

**Review by idlira — COMMENTED** [review]
I think the inccounter are translated on display.

File: src/main/java/sirius/biz/process/ErrorContext.java (hunk 2 above)
Comment (andyHa) [reply src/main/java/sirius/biz/process/ErrorContext.java:130]: If it is handled, our newly created exception would be ignored and our context would be lost..

File: src/main/java/sirius/biz/process/ErrorContext.java (hunk 1 above)
Comment (andyHa) [inline src/main/java/sirius/biz/process/ErrorContext.java:96]:
```suggestion
        withContext(label, value);
        try {
           task.execute();
        } finally {
           removeContext(label);
        }
```

**Review by ymo-sci — APPROVED** [review]

File: src/main/java/sirius/biz/process/ErrorContext.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+ * When importing data (especially for complex files like XML), the error messages generated by <tt>sirius-db</tt>
+ * are often not concise enough. Therefore, we populate this error context to keep track which entity was modified
+ * and which actual action as been attempted.
+ */
+public class ErrorContext implements SubContext {
+
+    /**
+     * Marks an error as enhanced.
+     * <p>
+     * We need to perform this, as most probably several layers wrap an <tt>perform</tt> block around
```
Comment (cschierle) [inline src/main/java/sirius/biz/process/ErrorContext.java:39]:
```suggestion
     * We need to perform this, as most probably several layers wrap a <tt>perform</tt> block around
```

File: src/main/java/sirius/biz/process/ErrorContext.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    /**
+     * Adds the given value to the context, performs the given task and then removes the value again.
+     *
+     * @param label the name of the value (this will be {@link NLS#smartGet(String) auto translated}.
+     * @param value the value to store
+     * @param task  the task to perform while the value is set
+     * @return the context for fluent method calls
+     * @throws Exception any error as thrown by the given task. Note that the exceptions and not handlend and also
```
Comment (cschierle) [inline src/main/java/sirius/biz/process/ErrorContext.java:90]:
```suggestion
     * @throws Exception any error as thrown by the given task. Note that the exceptions are not handled and also
```

File: src/main/java/sirius/biz/process/ErrorContext.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        String contextAsString = getContextAsString();
+        if (Strings.isFilled(contextAsString)) {
+            return errorMessage + " (" + contextAsString + ")";
+        } else {
+            return errorMessage;
+        }
+    }
+
+    /**
+     * Performs the given task and directly reports any error which occurs.
```
Comment (cschierle) [inline src/main/java/sirius/biz/process/ErrorContext.java:166]:
```suggestion
     * Performs the given task and directly reports any occurring error.
```

**Review by cschierle — APPROVED** [review]

**Review by qw3ry — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Pull request merged**